### PR TITLE
Working example with adbc-driver-snowflake

### PR DESCRIPTION
**Pull request by natemurthy**
Are there any Snowflake server/client compatibility or runtime version dependencies I should be aware of? For my team's specific data warehouse solution and code, I could only get this working with the `adbc-driver-snowflake` engine.

When I use the default `connectorx`, I get the following error every time:

```
  File "/Users/nathan/code/src/github.com/REDACTED/scripts/examples/polars_snowflake.py", line 52, in <module>
    df = pl.read_database(query, connection=cxn_str, engine="connectorx")
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/nathan/.pyenv/versions/REDACTED/lib/python3.11/site-packages/polars/utils/deprecation.py", line 93, in wrapper
    return function(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/nathan/.pyenv/versions/REDACTED/lib/python3.11/site-packages/polars/io/database.py", line 122, in read_database
    return _read_sql_connectorx(
           ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/nathan/.pyenv/versions/vREDACTED/lib/python3.11/site-packages/polars/io/database.py", line 153, in _read_sql_connectorx
    tbl = cx.read_sql(
          ^^^^^^^^^^^^
  File "/Users/nathan/.pyenv/versions/REDACTED/lib/python3.11/site-packages/connectorx/__init__.py", line 386, in read_sql
    result = _read_sql(
             ^^^^^^^^^^
RuntimeError: Source Unknown not supported.
```

### CODE REVIEW

**Comment by IndexSeek** [comment]
From what I know, there shouldn't be any compatibility or runtime dependencies to use Snowflake here. The only potential issue I can think of that might prevent you from connecting to Snowflake could be if there was an Authentication Policy in place to disallow connecting with a certain driver. However, the error message you are showing does not indicate that this is the case.

As you mentioned, I was able to get this working with the ADBC driver. This works.

```py
pl.read_database_uri(
    query="SELECT * FROM DATA.DIAMONDS",
    uri=URL(**connection_params),
    engine="adbc",
).head()
```

But then this does fail with the Runtime Error you were seeing:
```py
pl.read_database_uri(
    query="SELECT * FROM DATA.DIAMONDS",
    uri=URL(**connection_params),
    engine="connectorx",
).head()
```

But this does work:
```py
pl.read_database(
    "SELECT * FROM DATA.DIAMONDS", create_engine(URL(**connection_params)).connect()
)
```

I am curious, `read_database` does not support the engine argument from what I am finding, what happens if you exclude that? Here's what happened when I tried to include it:

```
ValueError                                Traceback (most recent call last)
Cell In[5], [line 1](vscode-notebook-cell:?execution_count=5&line=1)
----> [1](vscode-notebook-cell:?execution_count=5&line=1) pl.read_database(
      [2](vscode-notebook-cell:?execution_count=5&line=2)     "SELECT * FROM DATA.DIAMONDS", create_engine(URL(**connection_params)).connect(), engine="connectorx"
      [3](vscode-notebook-cell:?execution_count=5&line=3) )

File snowflake-to-polars/.venv/lib/python3.11/site-packages/polars/io/database/functions.py:271, in read_database(query, connection, iter_batches, batch_size, schema_overrides, infer_schema_length, execute_options, **kwargs)
    [269](https://file+.vscode-resource.vscode-cdn.net/snowflake-to-polars/snowflake-to-polars/.venv/lib/python3.11/site-packages/polars/io/database/functions.py:269) if kwargs:
    [270](https://file+.vscode-resource.vscode-cdn.net/snowflake-to-polars/snowflake-to-polars/.venv/lib/python3.11/site-packages/polars/io/database/functions.py:270)     msg = f"`read_database` **kwargs only exist for passthrough to `read_database_uri`: found {kwargs!r}"
--> [271](https://file+.vscode-resource.vscode-cdn.net/snowflake-to-polars/snowflake-to-polars/.venv/lib/python3.11/site-packages/polars/io/database/functions.py:271)     raise ValueError(msg)
    [273](https://file+.vscode-resource.vscode-cdn.net/snowflake-to-polars/snowflake-to-polars/.venv/lib/python3.11/site-packages/polars/io/database/functions.py:273) # return frame from arbitrary connections using the executor abstraction
    [274](https://file+.vscode-resource.vscode-cdn.net/snowflake-to-polars/snowflake-to-polars/.venv/lib/python3.11/site-packages/polars/io/database/functions.py:274) with ConnectionExecutor(connection) as cx:

ValueError: `read_database` **kwargs only exist for passthrough to `read_database_uri`: found {'engine': 'connectorx'}
```

I hope that we can get to the bottom of this! If now, we might need to raise an issue in https://github.com/pola-rs/polars. That  `RuntimeError: Source Unknown not supported.` error is peculiar.

**Comment by sfc-gh-kmason** [comment]
Hey @natemurthy, @IndexSeek and I did some relatively thorough testing with different version and connection combinations. Here are the results. 

We created a .py script that tries different connects and a shell script to sequentially create environments with different versions of polars. 

Here is the .py script:
```py
import os
import polars as pl
from dotenv import load_dotenv
from sqlalchemy import create_engine
from snowflake.connector import connect
import warnings
warnings.filterwarnings("ignore")

load_dotenv(override=True)

print("\npolars version: " + pl.__version__)

# URI
uri = os.environ.get("SNOWFLAKE_URI")

# SQLAlchemy Connection - produces sqlalchemy.engine.base.Connection 
sa_con = create_engine(os.environ.get("SNOWFLAKE_URI")).connect()

# Snowflake Connection (assumes a config.toml file exists) - produces snowflake.connector.connection.SnowflakeConnection
sf_con = connect(connection_name='default')

# Query
query = "SELECT * FROM ANALYTICS.SHIFT_SALES_TEST"

for con in (uri, sa_con, sf_con):
    # Test read_database
    try:
        df = pl.read_database(query, con)
        print("read_database", type(con), "SUCCESS")
    except Exception as e:
        print("read_database", type(con), "FAILED", e)

    # Test read_database_uri
    if isinstance(con, str):
        for engine in ('adbc', 'connectorx'):  
            try:
                df = pl.read_database_uri(query, con, engine=engine)
                print("read_database_uri", type(con), "engine=", engine, "SUCCESS")
            except Exception as e:
                print("read_database_uri", type(con), "engine=", engine, "FAILED", e) 

print("Done testing polars version " + pl.__version__ + "\n")
sa_con.close()
sf_con.close()
```

Here is the shell script:
```
uv venv --python 3.10
source .venv/bin/activate
uv pip install -r requirements.txt

versions=("0.18.15" "0.20.31" "1.0.0")
for version in "${versions[@]}"; do
    uv pip install "polars==$version"
    python polars_run.py
done
```

And here is the new requirements.txt file:
```
adbc-driver-snowflake==1.0.0
connectorx==0.3.3
notebook==7.2.1
pandas==2.2.2
pyarrow==16.1.0
python-dotenv==1.0.1
snowflake-sqlalchemy==1.5.3
sqlalchemy==1.4.52
```

Here is the output of executing the shell script:
```
Resolved 1 package in 3ms
Installed 1 package in 4ms
 + polars==0.18.15

polars version: 0.18.15
read_database <class 'str'> FAILED Source Unknown not supported.
read_database_uri <class 'str'> engine= adbc FAILED module 'polars' has no attribute 'read_database_uri'
read_database_uri <class 'str'> engine= connectorx FAILED module 'polars' has no attribute 'read_database_uri'
read_database <class 'sqlalchemy.engine.base.Connection'> FAILED Expect connection to be a URI string; found <class 'sqlalchemy.engine.base.Connection'>
read_database <class 'snowflake.connector.connection.SnowflakeConnection'> FAILED Expect connection to be a URI string; found <class 'snowflake.connector.connection.SnowflakeConnection'>
Done testing polars version 0.18.15

Resolved 1 package in 9ms
Installed 1 package in 7ms
 - polars==0.18.15
 + polars==0.20.31

polars version: 0.20.31
read_database <class 'str'> FAILED Source Unknown not supported.
read_database_uri <class 'str'> engine= adbc SUCCESS
read_database_uri <class 'str'> engine= connectorx FAILED Source Unknown not supported.
read_database <class 'sqlalchemy.engine.base.Connection'> SUCCESS
read_database <class 'snowflake.connector.connection.SnowflakeConnection'> SUCCESS
Done testing polars version 0.20.31

Resolved 1 package in 9ms
Installed 1 package in 6ms
 - polars==0.20.31
 + polars==1.0.0

polars version: 1.0.0
read_database <class 'str'> FAILED Use of string URI is invalid here; call `read_database_uri` instead
read_database_uri <class 'str'> engine= adbc SUCCESS
read_database_uri <class 'str'> engine= connectorx FAILED Source Unknown not supported.
read_database <class 'sqlalchemy.engine.base.Connection'> FAILED cannot import name 'async_sessionmaker' from 'sqlalchemy.ext.asyncio' (/snowflake-to-polars/.venv/lib/python3.10/site-packages/sqlalchemy/ext/asyncio/__init__.py)
read_database <class 'snowflake.connector.connection.SnowflakeConnection'> SUCCESS
Done testing polars version 1.0.0
```

In summary, we found the connectorx did not explicitly work. We also determined that the engine argument with the `read_database` method only works in 0.18 and deprecation warning is returned with version 0.20. Hope this helps.

**Comment by natemurthy** [comment]
@sfc-gh-kmason @IndexSeek thank you the dev support on this! The script immediately above is very helpful. I have a clear path forward on this, much appreciated!

Feel free to close this PR or update with these scripts :)

**Comment by natemurthy** [comment]
@sfc-gh-kmason @IndexSeek thank you the dev support on this! The script immediately above is very helpful. I noticed you're using virtual environment with Python 3.10. My team is currently on Python 3.11 (with some of us on Python 3.12). I'll run the above test on the other interpreter versions, but seeing if you're aware of any runtime compatibility expectations?

Either way, I have a clear path forward on this, much appreciated! Feel free to close this PR or update with these scripts :)
